### PR TITLE
Add %ssl::<cert macro for logging server X.509 certificate

### PR DESCRIPTION
**Pull request by redmeadowman**
We have chosen the PEM format instead of, for example, raw DER format
because most programs exchange certificates using PEM format and because
logging raw binary values would be unusual for Squid logformat.

The current support is limited to SslBump step3 which parses and stores
the peer certificate. TODO: Support all from-Squid TLS connections.

### CODE REVIEW

**Comment by yadij** [comment]
Please document in the description what use this is, and use 72 or less characters as line wrapping.

FYI: The multi-line behaviour will be very problematic. logformat macros are used in single-line log entries, single-line helper queries, and single-line URLs - and Squid logformat are required to support multiple different quotings. The only multi-line uses are for the cert generator and validator helpers which are already being passed the certificate in PEM format.

**Comment by redmeadowman** [comment]
I've made the changes suggested by rousskov, namely, changed the logformat command to ssl::<cert, and added the more detailed documentations.  The behavior I see during testing matches the documentation description provided by rousskov.  The modifications have been committed, squashed, and pushed.

**Comment by rousskov** [comment]
FYI: There is no need to manually squash your PR changes. [Anubis](https://wiki.squid-cache.org/MergeProcedure#Automation) will do that for you when it comes to merging the PR. Squashing did not hurt now, but, in general, squashing may negatively affect reviews and GitHub change requests.

**Comment by redmeadowman** [comment]
Thanks rousskov for the tip on not needing to squash commits.  FYI, squashing commits is actually spelled out in the Squid Merge Procedure document, under "Pull Request", item 5, so we may want to request the maintainer of that document to remove the requirement.
https://wiki.squid-cache.org/MergeProcedure

**Comment by rousskov** [comment]
> we may want to request the maintainer of that document to remove the requirement

I have updated that requirement to reflect the current status. That document, like 99% of the Squid wiki, is maintained by the community. Feel free to fix any bugs that you are comfortable fixing there!

**Comment by rousskov** [comment]
> sslGetUserCertificatePEM() stores the resulting PEM string into a static variable. Ssl::GetX509PEM() does the same, so keeping these two static variables separate unique is important.

I do not see why sslGetUserCertificatePEM() cannot return Ssl::GetX509PEM() result _instead_ of using a local static variable. Such code reuse is better than duplicating that ugly code IMO.

Ideally, neither function should abuse static locals for results. Both functions should return an SBuf instead. I have added a change request about that conversion, but (as I have already said in that change request), I do not insist on those changes. They would make function reuse obviously safe, but that reuse is still possible (and is a good idea) without SBuf conversion.

**Comment by redmeadowman** [comment]
Alex,
I believe the most recent changes I submitted cover all the items you requested.  I've compile and tested, and the new functionality works as expected.

**Comment by rousskov** [comment]
Thank you very much for polishing this up! I see a couple of spots that were either missed or can be easily improved. It may be easier/faster to fix them than to do another review iteration. Do you mind if I commit those changes into your PR branch? Normally, I would do that without asking (where authors has permitted such modifications when opening the PR), but you are using _master_ instead of a dedicated feature branch for this PR, and I do not want to modify your master without your explicit permission.

**Comment by redmeadowman** [comment]
Alex, feel free to make any cleanup changes you see necessary.

File: src/cf.data.pre [hunk 1]
```diff
@@ -4765,6 +4765,26 @@ DOC_START
...
+
+				Currently, Squid only collects server
+				certificates during step3 of SslBump
+				processing; connections that were not subject
+				to ssl_bump rules or that did not match a peek
+				or stare rule at step2 will not have the
+				server certificate information.
+
+				The default encoding for this field is
+				multi-line. When Squid cannot log a
```
Comment (rousskov) [inline src/cf.data.pre]: I think the default encoding should not be multi-line because we should not break access.log line-based format by _default_ and because a "quoted string" encoding is easy to support (by adding `quote = 1` to the relevant Format.cc code IIRC). I have proposed using quoted string encoding before, but you insisted on using multi-line default. Why do you insist on it?

Comment (redmeadowman) [reply src/cf.data.pre]: Hi Alex,

Sorry if I made it seem as though I insisted on multi-line as the default.  I simply observed that multi-line was the default the way I copied and pasted it.  Please feel free to make the default whichever way you see best.

Thanks for accepting this pull request.

**Comment by redmeadowman** [comment]
My testing reveals that the certificate output values appears to be truncated after around 8000 characters (depending if you count carriage returns or not).  Are there inherit limits built into the logging subsystem that could cause this?

**Comment by rousskov** [comment]
Glad you caught this truncation problem before we committed this feature! It should be addressed to avoid adding field truncation bugs like the one recently fixed in bec110e -- individual log fields should not be truncated by default. If the field does not fit, Squid should log a dash `-` (with a one-time cache.log warning detailing the problem).

It should be easy for you to find the source of truncation by stepping through major certificate conversion steps in a debugger. IIRC, the log lines (as a whole) do not have an 8KB limit, so I would expect the truncation to happen either in certificate handling code or perhaps in generic encoding code.

**Comment by redmeadowman** [comment]
Alex, can you direct me to code in squid that prints the "-" and a warning to cache.log in a failure case?

**Comment by rousskov** [comment]
@dnevil, I do not think it is a good idea to combine the primary changes in this PR (i.e. a new logformat field) with the changes required to increase total access.log line limits (i.e. your fcf7c27). The latter change is probably controversial and may require a lot of work to get it right (e.g., your current version is not good enough). Moreover, the fcf7c27 approach does not fully solve the problem -- a certificate might still be larger than the new 1MB line limit.

If there is no PR code that limits the size of the certificate PEM representation, and the only limit you are facing is the total access.log line limit, then I would not try to solve both problems in one PR. IMO, it would even be OK to merge this PR first if many certificates can be successfully logged without line limit changes; we just would need to add a warning about the total line limit problem to the new logformat %code description in cf.data.pre.

In summary, the new certificate logging code should not truncate PEM values, but removing or increasing total access.log line limits is outside this PR scope IMO, even if such removal or increase is necessary to log many PEM values.

**Comment by redmeadowman** [comment]
Agreed, I will move the log limit code to a new pull request.

File: src/cf.data.pre (hunk 1 above)
Comment (redmeadowman) [reply src/cf.data.pre]: The default encoding text was removed from the documentation file cf.data.pre.

**Comment by yadij** [comment]
I agree with Alex that this is not a clear benefit until that line-limit issue is resolved, but we do not have a clear case to reject either since the limits are a site-specific problem.

So, please add a "WARNING: " prefixed note to the cf.data.pre text stating that some certificates are large and may cause issues by exceeding limits which differ by log module used. Such that any admin considering this code/tag will not hit the side effects unaware. I will mark it as passed for merge after that addition and let our slow-path (no positive votes) process take over.

If someone were to use this as a foundation for logging arbitrary certificate fields/attributes with the {arg} parameter it would have a much stronger case for inclusion and get my vote. But that is possibly a fair bit of work to add, so I do not insist on that as part of this PR.

**Comment by rousskov** [comment]
> I will mark it as passed for merge after that addition and let our slow-path (no positive votes) process take over.

There is no such process(**). The "slow burner" path still requires a positive vote.

However, if/when you are happy with this PR being merged but do not want to approve it for some reason, then I can approve this PR so that it is merged normally. In other words, I think this specific PR needs two positive votes (or equivalent), but I do not want to abuse the `M-cleared-for-merge` mechanism to emulate that. The note below explains why using that label for this purpose should be considered abuse.

----

(**) To be more precise, Anubis does not support unapproved merges today and may never support them because that process is essentially incompatible with GitHub (and similar) PRs -- votes is what gets PRs merged. Yes, we have the `M-cleared-for-merge` label as a separate merge control step right now, but it is currently designed as a _temporary_ measure on _top_ of approval voting, to be disabled when we trust Anubis to do the right thing.

From the UX point of view, having two very different merge approval layers is an inferior design. As we get more experience and discover/document our evolving PR merge needs, if needed, instead of disabling `M-cleared-for-merge`, we can certainly redesign merge control so that there is only one layer with more nuanced votes/roles, but it will require a lot of work and is unlikely to happen in the foreseeable future.

For now, we should use votes to control merging and use the `M-cleared-for-merge` label only as a temporary "we do not fully trust" Anubis feature.

**Comment by yadij** [comment]
>Amos:
>    I will mark it as passed for merge after that addition and let our slow-path (no positive votes) process take over.

> Alex:
> There is no such process(**). The "slow burner" path still requires a positive vote.

I am referring to our published merge procedure clause 4: "Submissions older than 10 days without negative votes are accepted.". But as you say, Anubis currently needs manual confirmation to apply submissions. So in practice that means 10+ days after the final tweaks are finalized it will get my "vote" and labeling for merge unless there is a reason found to apply a negative vote before that.

**Comment by rousskov** [comment]
> I am referring to our published merge procedure clause 4: "Submissions older than 10 days without negative votes are accepted."

I know. That [publication](https://wiki.squid-cache.org/MergeProcedure#Votes), including its [exceptions](https://wiki.squid-cache.org/MergeProcedure#Exceptions), is mostly obsolete. It does not accurately reflect the current reality (also published and agreed on!).

Specifically, I believe that clause 4 must be dropped. It is not automated, and we should not manually emulate it.

Historically, clause 4 was not accurately followed -- I doubt any committer merged a patch just because it has been posted for 10 days! Our GitHub experience so far suggests that the clause is not essential for making progress. I do not know of any serious project, including serious open source projects that honestly uses a similar clause. That clause can be abused to commit controversial changes without taking the responsibility for those commits. Its only positive effect (that I can think of) is that it creates an impression of a project democracy, but that impression is just a mirage (of what is actually a bad idea in a software development context).

Instead, each merge should be meaningfully reviewed and must be carefully approved by a core developer. The existing automation already covers that procedure pretty well.

If you agree, I can volunteer to adjust the wiki accordingly or, if you prefer, post an adjusted version for a review on squid-dev.

**Comment by rousskov** [comment]
> "just" no

Exactly. The rest does not matter. The committer who merges the patch should review it and must approve it. And we already have mechanisms for doing both. Clause 4 is wrong, misleading, and unnecessary.

> See PR #77 is a good example of a single committer merging via the delayed process without additional reviews.

That PR was not merged using clause 4. It was merged because a committer reviewed and approved it. Exactly the same thing would have happened if clause 4 did not exist.

> Changing the conditions imposed on contributors to Squid is a board decision

It is a Squid Project decision. The Board may get involved, especially if there are conflicts that need to be resolved, of course, but the Board is not the only stakeholder in this decision. I could even argue that this decision has been made already. However, I will waste my time to post a proposal since you seem to insist on that.
